### PR TITLE
[WIP] new lineWidget upgrade 

### PR DESCRIPTION
**Pull request by ocrossi**
Add new functionalities to line widget including:
- better example
- new handle shapes
- better visibility feature
- handle rotations to face camera
- added documentation

### CODE REVIEW

**Comment by ocrossi** [comment]
@finetjul i think we are good to go
Can you confirm [here](https://github.com/ocrossi/vtk-js/blob/314f230b0efa9f8d5d4a71d774eed6536755635d/Sources/Filters/Sources/CircleSource/index.js#L51) that it is the change you wanted in CircleSource? I'm not sure about what i did there

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 1]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+    const visibility = handle.getVisible();
+    model.representations[nb - 1].setVisibilityFlagArray([
+      visibility,
+      handle.getShape() !== 'voidSphere',
+    ]);
+    handle.setVisible(visibility);
+    model.representations[nb - 1].updateActorVisibility();
+    if (nb === 1) {
+      model.representations[nb].updateActorVisibility();
+    }
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:247]: Is the following the same ?
```
model.representations.filter((r, index) => index <=nb).forEach((r) -> r->updateActorVisibility)
```

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 2]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+    model.interactor.render();
+  };
+
+  // --------------------------------------------------------------------------
+
+  publicAPI.placeHandle = (handle) => {
+    const nb = getNbHandles(model);
+    publicAPI.manageHandleVisibility(handle, nb);
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:255]: it seems to me that nb could be computed inside `manageHandleVisibility` instead

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 3]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+    model.nbHandles = nb;
+    handle.setOrigin(...model.widgetState.getMoveHandle().getOrigin());
+    handle.setColor(model.widgetState.getMoveHandle().getColor());
+    handle.setScale1(model.widgetState.getMoveHandle().getScale1());
+    model.widgetState.getText().setText(model.text);
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:260]: why overwrite widget state text ?

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 4]
```diff
@@ -187,30 +316,38 @@ export default function widgetBehavior(publicAPI, model) {
...
+          model.nbHandles === 1 &&
+          isHandleOrientable(handle1.getShape())
         ) {
-          updateHandleDirection(HandleBehavior.HANDLE1_ALONE, callData);
+          const direction = orientFirstHandleBeforeSecondHandlePlacement(
+            callData
+          );
+          orientHandle(1, direction);
+        }
+        if (model.nbHandles === 1 && getNbHandles(model) === 2) {
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:341]: you've got to add a comment to explain this strange line

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 5]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+    return dir;
+  }
+
+  function orientHandle(nb, direction) {
+    if (getNbHandles(model) !== 2) {
+      return;
     }
-    if (isHandleOrientable(model.handle2Shape)) {
-      updateHandleDirection(HandleBehavior.HANDLE2);
+    console.log(getNbHandles(model));
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:155]: to remove

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 6]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+
+    if (
+      (nb === 1 && handle1.getShape() === 'cone') ||
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:158]: use enums instead of directly their values ('cone' -> `???.CONE`)

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 7]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+        ? handle2.getOrigin()
+        : model.widgetState.getMoveHandle().getOrigin();
+    dir = computeDirectionWithTwoPoints(handle1Origin, handle2Origin);
+    if (nb === 2 || nb === 3) {
+      vtkMath.multiplyScalar(dir, -1);
+    }
+    return dir;
+  }
+
+  function orientHandle(nb, direction) {
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:151]: this `nb` parameter is strange. 
Please add function documentation.
It's not easy to read, because you update the handle `nb-1` and not `nb`.
It seems to me that it is more a `handleIndex` that this function must take as parameter rather than a pseudo "number of handles"  (which is different from `getNbHandles`)

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js [hunk 8]
```diff
@@ -30,85 +27,240 @@ export default function widgetBehavior(publicAPI, model) {
...
+    }
+    if (
+      (model.representations[1].getFaceCamera() === FaceCameraStates.NONE &&
+        publicAPI.isHandleGlyph2D(handle2.getShape())) ||
+      model.representations[1].getFaceCamera() === FaceCameraStates.TRUE
+    ) {
+      model.representations[1].setViewMatrix(
+        Array.from(model.camera.getViewMatrix())
+      );
+    }
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/behavior.js:199]: I wouldn't bother doing the test here (wether FaceCamera is forced or not), I would let the representation deal with it.

Here you simply always set the view matrix.

File: Sources/Widgets/Widgets3D/LineWidget/helper.js [hunk 9]
```diff
@@ -16,3 +18,25 @@ export function updateTextPosition(model) {
   const SVGTextState = model.widgetState.getText();
   SVGTextState.setOrigin(calculateTextPosition(model));
 }
+
+/* *
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/helper.js:22]: /**

File: Sources/Widgets/Widgets3D/LineWidget/helper.js [hunk 10]
```diff
@@ -16,3 +18,25 @@ export function updateTextPosition(model) {
...
+ * Returns the number of handle placed on the scene by checking
+ * handle positions. Returns 2 when the user is still
+ * placing 2nd handle
+ * */
+
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/helper.js:27]: no empty line

File: Sources/Widgets/Widgets3D/LineWidget/helper.js [hunk 11]
```diff
@@ -16,3 +18,25 @@ export function updateTextPosition(model) {
...
+export function getNbHandles(model) {
+  const handle1 = model.widgetState.getHandle1();
+  const handle2 = model.widgetState.getHandle2();
+  const moveHandle = model.widgetState.getMoveHandle();
+  if (handle1.getOrigin().length !== 0 && handle2.getOrigin().length !== 0) {
+    return 2;
+  }
+  if (
+    vtkMath.areEquals(handle1.getOrigin(), handle2.getOrigin(), 0) ||
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/helper.js:36]: add a comment to explain why they are equal

File: Sources/Widgets/Widgets3D/LineWidget/index.js [hunk 12]
```diff
@@ -1,88 +1,36 @@
...
 
-const { HandleRepresentationType, HandleRepresentation } = Constants;
-
-const shapeToRepresentation = {};
-
 function vtkLineWidget(publicAPI, model) {
   model.classHierarchy.push('vtkLineWidget');
+  model.widgetState = stateGenerator();
+  model.behavior = widgetBehavior;
+  model.handleRepresentations = [0, 0];
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/index.js:23]: why 0 ?

Comment (finetjul) [reply Sources/Widgets/Widgets3D/LineWidget/index.js:23]: do you need this variable anyway?

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js (hunk 2 above)
Comment (finetjul) [reply Sources/Widgets/Widgets3D/LineWidget/behavior.js:255]: I would set visibility before configuring the handles

File: Sources/Widgets/Widgets3D/LineWidget/state.js [hunk 13]
```diff
@@ -1,36 +1,46 @@
...
     .addStateFromMixin({
       labels: ['handle1'],
-      mixins: ['origin', 'color', 'scale1', 'visible'],
+      mixins: ['origin', 'color', 'scale1', 'visible', 'manipulator', 'shape'],
       name: 'handle1',
       initialValues: {
         scale1: 50,
         origin: [],
-        visible: false,
+        // shape: 'cone',
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/state.js:33]: to remove

File: Sources/Widgets/Widgets3D/LineWidget/state.js [hunk 14]
```diff
@@ -1,36 +1,46 @@
...
     .addStateFromMixin({
       labels: ['handle2'],
-      mixins: ['origin', 'color', 'scale1', 'visible'],
+      mixins: ['origin', 'color', 'scale1', 'visible', 'manipulator', 'shape'],
       name: 'handle2',
       initialValues: {
         scale1: 50,
         origin: [],
-        visible: false,
+        // shape: 'cone',
```
Comment (finetjul) [inline Sources/Widgets/Widgets3D/LineWidget/state.js:43]: to remove

**Comment by finetjul** [comment]
> @finetjul i think we are good to go
> Can you confirm [here](https://github.com/ocrossi/vtk-js/blob/314f230b0efa9f8d5d4a71d774eed6536755635d/Sources/Filters/Sources/CircleSource/index.js#L51) that it is the change you wanted in CircleSource? I'm not sure about what i did there

Yes for CircleSource

File: Sources/Filters/Sources/Arrow2DSource/index.js [hunk 15]
```diff
@@ -151,7 +155,6 @@ function vtkTriangleSource(publicAPI, model) {
   vtkMatrixBuilder
     .buildFromRadian()
     .translate(...model.origin)
-    .rotateFromDirections([0, 1, 0], model.direction)
```
[the remark this answers is not in the document]
Comment (ocrossi) [reply Sources/Filters/Sources/Arrow2DSource/index.js:154]: the rotation is managed in arrow2DHandleRepresentation so no need to do it at the glyph source level.

File: Sources/Filters/Sources/CircleSource/index.js [hunk 16]
```diff
@@ -62,6 +71,7 @@ const DEFAULT_VALUES = {
   radius: 1.0,
   resolution: 6,
   center: [0, 0, 0],
+  orientation: [1, 0, 0],
```
[the remark this answers is not in the document]
Comment (ocrossi) [reply Sources/Filters/Sources/CircleSource/index.js:74]: On 1st iterations of lineWidget the purpose was to 'direct' the arrow in the polyLine axis. Then came the feature of handle rotation to face camera, thus the different names.

File: Sources/Filters/Sources/Arrow2DSource/index.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Filters/Sources/Arrow2DSource/index.js:154]: Let's think big, maybe that souce will be used by something other than `arrow2HandleRepresentation`.
You can keep that feature to specify a given direction, it does not hurt :-)

File: Sources/Filters/Sources/CircleSource/index.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Filters/Sources/CircleSource/index.js:74]: Ok, let me rephrase: please use the same variable name between CircleSource and Arrow2DSource :-)

File: Sources/Widgets/Widgets3D/LineWidget/api.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    - Disk
+    - Circle
+    - ViewFinder
+
+## Handle orientation
+
+### Shape Direction
+Shapes meant to represent an arrow will follow the direction of the line. Those shapes include cone, arrow head (4 and 6) and triangle.
+
+### Shape Rotation
```
[the remark this answers is not in the document]
Comment (ocrossi) [reply Sources/Widgets/Widgets3D/LineWidget/api.md:32]: rotation and direction are 2 separate features to me. One is meant to match handle direction with polyline (to have a straight arrow) and the other is to rotate handle to face camera. Should i still change it?

File: Sources/Widgets/Representations/ArrowHandleRepresentation/index.js [hunk 18]
```diff
@@ -176,16 +352,24 @@ function vtkArrowHandleRepresentation(publicAPI, model) {
...
+  faceCamera: FaceCameraStates.NONE,
 };
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
   Object.assign(model, DEFAULT_VALUES, initialValues);
 
+  model.viewMatrix = mat4.create();
+  model.orientation = [1, 0, 0];
```
[the remark this answers is not in the document]
Comment (ocrossi) [reply Sources/Widgets/Representations/ArrowHandleRepresentation/index.js:365]: Why? I dont understand your comment

File: Sources/Widgets/Widgets3D/LineWidget/api.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Widgets/Widgets3D/LineWidget/api.md:32]: I was only picking on on the uppercase for the second word. nothing more

File: Sources/Widgets/Representations/ArrowHandleRepresentation/index.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Widgets/Representations/ArrowHandleRepresentation/index.js:365]:
```
model.viewMatrix = mat4.create();
model.orientation = [1, 0, 0];
Object.assign(model, DEFAULT_VALUES, initialValues);
```
and
```
Object.assign(model, DEFAULT_VALUES, initialValues);
model.viewMatrix = mat4.create();
model.orientation = [1, 0, 0];
```
is not the same. Which one makes more sense to you ?

[the remark this answers is not in the document]
Comment (ocrossi) [reply Sources/Widgets/Representations/ArrowHandleRepresentation/index.js:365]: On first look i would say it is the same, as assigning initialValues should not prevent us from adding properties in this object. Upon reflexion i would choose 1st solution if the viewMatrix and orientation get somehow overriden by initialValues

**Comment by jourdain** [comment]
In general, we should never use `mat4.create()` as those are float32 rather than float64... It is better to create float64 ourself...

**Comment by laurennlam** [comment]
> In general, we should never use `mat4.create()` as those are float32 rather than float64... It is better to create float64 ourself...

Indeed... we had an issue with the reslice cursor because of it... (see #1747)

File: Sources/Widgets/Representations/ArrowHandleRepresentation/index.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (finetjul) [reply Sources/Widgets/Representations/ArrowHandleRepresentation/index.js:365]: initialValues should overwrite default values. 
This is precisely why it is `Object.assign(model, DEFAULT_VALUES, initialValues)` and not `Object.assign(model, initialValues, DEFAULT_VALUES);`
So you're right, 1st solution is the way to go. Alternatively, you can see https://github.com/Kitware/vtk-js/issues/1751

File: Sources/Widgets/Widgets3D/LineWidget/behavior.js (hunk 2 above)
Comment (ocrossi) [reply Sources/Widgets/Widgets3D/LineWidget/behavior.js:255]: this is what i am doing in my mind. 1st thing to be called after getting handle number data (which handle is placed) is setting visibility feature. Am i missing something here?

Comment (finetjul) [reply Sources/Widgets/Widgets3D/LineWidget/behavior.js:255]: set origin color, scale and text BEFORE calling manageHandleVisibility

File: Sources/Interaction/Widgets/ResliceCursor/ResliceCursorRepresentation/index.js [hunk 19]
```diff
@@ -210,19 +210,20 @@ function vtkResliceCursorRepresentation(publicAPI, model) {
...
     vtkMath.normalize(viewUp);
     model.viewUpFromViewType[planeNormalType] = viewUp;
 
+    transformPlane(
+      model.planeSource,
+      publicAPI.getResliceCursor().getCenter(),
+      plane.getNormal(),
+      viewUp
+    );
+
```
Comment (finetjul) [inline Sources/Interaction/Widgets/ResliceCursor/ResliceCursorRepresentation/index.js:226]: Please move this commit in a separate MR

**Comment by ocrossi** [comment]
still in progress, see #1759
